### DISCUSSION TITLE
[MacOS] window-inherit-working-directory enabled but unexpectedly not working

### DISCUSSION THREAD

**Discussion by WizardOfMenlo** [ghostty-org/ghostty#4176] [Q&A] [answered]
Hi, I have been using Ghostty for a week now and it has worked like a charm.
Today, the window-inherit-working-directory feature has unexpectedly stopped working. 
Expected behaviour: I cd into a directory, press Cmd+D to open a tab, and that tab is in the same directory
Observed behaviour: the tab opens in the ~ directory instead.

**Config**
```
font-family = TX02 Nerd Font Mono
#font-family = FiraCode Nerd Font Mono Ret
font-size = 16
theme = GruvboxDarkHard
cursor-style = block
adjust-cell-height = 20%
mouse-hide-while-typing = true
mouse-scroll-multiplier = 2
macos-titlebar-style = hidden
window-height = 30
window-width = 100
window-padding-balance = true
window-save-state = always
window-colorspace = "display-p3"
background = 1C2021
window-inherit-working-directory = true
```
**Version**:
```
Ghostty 1.0.0

Version
  - version: 1.0.0
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.none
  - font engine: font.main.Backend.coretext
  - renderer   : renderer.Metal
  - libxev     : main.Backend.kqueue
```

**Comment by johnallen3d** [comment]
Oh good, I thought I must be the only one with this issue! I've been using Ghostty since release day and have always experienced this issue.

**Comment by grEvenX** [comment]
I'm not 100% sure, but it seems that in my case the problem was fixed when updating my oh-my-zsh setup (`omz update`). Though I saw the same problem when I didn't even load oh-my-zsh 🤷‍♂️ Maybe worth giving it a try if you are using it.

**Reply by johnallen3d** [reply]
hmm I'm using Fish shell. I forced my `shell-integration` to `fish` thinking that might help but it hasn't had an impact.

**Comment by aryounce** [comment]
@WizardOfMenlo and @johnallen3d are either of you using `/bin/bash` by chance? I've been researching `window-inherit-working-directory` as it doesn't work for me either and it appears [the necessary shell integration is disabled](https://github.com/ghostty-org/ghostty/blob/main/src/termio/shell_integration.zig#L63) for the built in `/bin/bash` on macOS.

I'm digging into it a bit more and will follow up.

**Reply by johnallen3d** [reply]
I use Fish and setting (or not setting) `shell-integration = fish` seems to have no impact on this feature for me.

**Comment by WizardOfMenlo** [comment]
@aryounce Hey, no I am using `/bin/zsh`. However, I added `shell-integration = zsh` in my config and now it seems to be back to working.

**Reply by WizardOfMenlo** [reply]
Update, it randomly stopped working again...

**Comment by will-stone** [comment]
I have found that <kbd>cmd+D</kbd> or <kbd>cmd+T</kbd> fails to open current directory if I've got a running Vitest dev process going (might be the same for other running processes? 🤷‍♂️). This is pretty much the only bug I've experienced in Ghostty so kudos to that! 😄

**Reply by will-stone** [reply]
My setup:
- Ghostty 1.0.1
- macOS 14.5
- Fish 3.7.1
- Tide prompt 6.1.1

**Comment by achwo** [comment]
For me when pwd contains a `?`, the new tab starts in `/`.
i.e. pwd: `~/test?` => new tab: `/`

I use Ghostty 1.0.1, macOS 15.2, zsh 5.9

**Comment by johnallen3d** [comment]
FWIW - I'm using Nix/`nix-darwin`/`nix-homebrew`/`home-manager` and the solution [here](https://github.com/ghostty-org/ghostty/discussions/4725#discussioncomment-11755031) worked for me (e.g. manually setting `XDG_DATA_DIRS`). After applying this change (and restarting Ghostty) this feature is working for me and I was able to remove the setting `shell-integration = fish` from my config (which apparently wasn't working anyway).

**Comment by luebken** [comment]
Just bumping this that the problem still persists for me.

ghostty --version
Ghostty 1.1.3

Version
  - version: 1.1.3
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.none
  - font engine: font.main.Backend.coretext
  - renderer   : renderer.Metal
  - libxev     : main.Backend.kqueue
  
config:
window-width = 110
window-height = 34
window-inherit-working-directory = true

**Reply by rhodes-b** [reply]
inherit working directory is part of shell integration which shell are you using

**Reply by 00-kat** [reply]
Please [ensure that shell integration actually worked](https://ghostty.org/docs/features/shell-integration#verify-shell-integration).  You can view logs on macOS with `sudo log stream --level debug --predicate 'subsystem=="com.mitchellh.ghostty"'`.

**Reply by paulgalow** [reply]
Thanks for sharing that command @00-kat. I'm running into the same issue. Using `window-inherit-working-directory = true` does not seem to have any effect. Shell integration seems to work based on the log output. What am I missing?

Environment:
```sh
$ ghostty --version
Ghostty 1.1.3

Version
  - version: 1.1.3
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.none
  - font engine: font.main.Backend.coretext
  - renderer   : renderer.Metal
  - libxev     : main.Backend.kqueue
```

Config:
```
font-family = SF Mono
font-size = 14
window-inherit-working-directory = true
window-save-state = always
window-colorspace = "display-p3"
confirm-close-surface = false
```

<details>
<summary>Display log stream output when starting Ghostty</summary>
<pre>
[com.mitchellh.ghostty:config] default shell src=passwd value=/bin/zsh
[com.mitchellh.ghostty:config] default working directory src=passwd value=/Users/<redacted>
[com.mitchellh.ghostty:AppDelegate] application will restore window state
[com.mitchellh.ghostty:ghostty] known but unimplemented action action=32
[com.mitchellh.ghostty:font_shared_grid_set] font regular: SF Mono Regular
[com.mitchellh.ghostty:font_shared_grid_set] font bold: SF Mono Bold
[com.mitchellh.ghostty:font_shared_grid_set] font italic: SF Mono Bold Italic
[com.mitchellh.ghostty:font_shared_grid_set] font bold_italic: SF Mono Bold Italic
[com.mitchellh.ghostty:io_exec] found Ghostty resources dir: /Applications/Ghostty.app/Contents/Resources/ghostty
[com.mitchellh.ghostty:io_exec] shell integration automatically injected shell=termio.shell_integration.Shell.zsh
[com.mitchellh.ghostty:ghostty] known but unimplemented action action=19
[com.mitchellh.ghostty:io_exec] started subcommand path=/usr/bin/login pid=37354
[com.mitchellh.ghostty:ghostty] config reload notification
[com.mitchellh.ghostty:ghostty] config change notification
[com.mitchellh.ghostty:ghostty] config reload notification
[com.mitchellh.ghostty:ghostty] config change notification
[com.mitchellh.ghostty:ghostty] config change notification
[com.mitchellh.ghostty:osc] invalid OSC command: 697;OSCUnlock=<redacted>
[com.mitchellh.ghostty:osc] invalid OSC command: 697;Dir=/Users/<redacted>
[com.mitchellh.ghostty:osc] invalid OSC command: 697;Shell=zsh
[com.mitchellh.ghostty:osc] invalid OSC command: 697;ShellPath=/bin/zsh
[com.mitchellh.ghostty:osc] invalid OSC command: 697;PID=37362
[com.mitchellh.ghostty:osc] invalid OSC command: 697;ExitCode=0
[com.mitchellh.ghostty:osc] invalid OSC command: 697;TTY=/dev/ttys001
[com.mitchellh.ghostty:osc] invalid OSC command: 697;Log=
[com.mitchellh.ghostty:osc] invalid OSC command: 697;ZshAutosuggestionColor=
[com.mitchellh.ghostty:osc] invalid OSC command: 697;FigAutosuggestionColor=
[com.mitchellh.ghostty:osc] invalid OSC command: 697;User=<redacted>
[com.mitchellh.ghostty:osc] invalid OSC command: 697;StartPrompt
[com.mitchellh.ghostty:osc] invalid OSC command: 697;EndPrompt
[com.mitchellh.ghostty:osc] invalid OSC command: 697;NewCmd=<redacted>
</pre>
</details>

**Comment by tmerse** [comment]
Happens to me after running npm scripts like `npm run build:watch` which is aliased to `tsc --watch --skipLibCheck`.

```
Ghostty 1.1.3

Version
  - version: 1.1.3
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.none
  - font engine: font.main.Backend.coretext
  - renderer   : renderer.Metal
  - libxev     : main.Backend.kqueue
```

- MaOs Sequoia 15.6
- fish shell

Here are the logs, for 

1. `npm run build:watch`
2. Kill process via `c-c`
3. Opening new split (which opens at `/`)

```
ghostty: [com.mitchellh.ghostty:stream] unimplemented OSC callback: terminal.osc.Command{ .end_of_command = terminal.osc.Command.Command__struct_20675{ .exit_code = 1 } }
ghostty: [com.mitchellh.ghostty:stream] unimplemented OSC command: end_of_command
ghostty: [com.mitchellh.ghostty:stream] unimplemented OSC callback: terminal.osc.Command{ .end_of_command = terminal.osc.Command.Command__struct_20675{ .exit_code = 1 } }
ghostty: [com.mitchellh.ghostty:stream] unimplemented OSC command: end_of_command
ghostty: [com.mitchellh.ghostty:stream] unimplemented OSC callback: terminal.osc.Command{ .end_of_command = terminal.osc.Command.Command__struct_20675{ .exit_code = null } }
ghostty: [com.mitchellh.ghostty:stream] unimplemented OSC command: end_of_command
ghostty: [com.mitchellh.ghostty:osc] unknown semantic prompts option: special_key
ghostty: [com.mitchellh.ghostty:ghostty] known but unimplemented action action=32
ghostty: [com.mitchellh.ghostty:io_exec] found Ghostty resources dir: /Applications/Ghostty.app/Contents/Resources/ghostty
ghostty: [com.mitchellh.ghostty:io_exec] shell integration automatically injected shell=termio.shell_integration.Shell.fish
ghostty: [com.mitchellh.ghostty:ghostty] known but unimplemented action action=19
ghostty: [com.mitchellh.ghostty:io_exec] started subcommand path=/usr/bin/login pid=80369
ghostty: [com.mitchellh.ghostty:osc] unknown semantic prompts option: special_key
ghostty: [com.mitchellh.ghostty:stream] unimplemented OSC callback: terminal.osc.Command{ .end_of_command = terminal.osc.Command.Command__struct_20675{ .exit_code = null } }
ghostty: [com.mitchellh.ghostty:stream] unimplemented OSC command: end_of_command
ghostty: [com.mitchellh.ghostty:osc] unknown semantic prompts option: special_key
```

Relevant config settings

```
working-directory = inherit
window-inherit-working-directory = true
shell-integration = fish
```

**Comment by didibus** [comment]
The issue for me was that since I have prompt hooks in my Zsh, like Starship, it messed up the shell injection.

I got it fixed by manually adding to my `.zshrc` before my prompt hooks the shell integration for Ghostty as found here: https://github.com/ghostty-org/ghostty/blob/main/src/shell-integration/README.md

**Reply by akriegman** [reply]
this solved it for me, thanks!

**Reply by paulgalow** [reply]
This did solve it for me and my Starship setup as well. Thanks so much for sharing.

**Comment by mitchellh** [comment, marked as answer]
I'm going to close this issue. It's become a catch-all of this simply not working, but not referencing any single actionable bug item and in most cases being user error in some form (Q&A is good for that). If there are specific issues, feel free to open a new discussion. :)
